### PR TITLE
Invoice CRUD

### LINKED ISSUE #13: US_invoices#index
User can see all invoices for a vehicle with basic info

### LINKED ISSUE #14: US_invoices#show
User can see the details of an invoice for one vehicle

### LINKED ISSUE #15: US_invoices#new-create
User can add an invoice for one vehicle

### LINKED ISSUE #16: US_invoices#edit-update
User can edit an invoice for one vehicle

### PR DESCRIPTION
**Pull request by JPS09**

### CODE REVIEW

**Review by Mordhin — APPROVED** [review]

**Comment by Mordhin** [comment]
C'est trop cool !

**Comment by sbriclot** [comment]
One important feature is missing.
When you add or edit an invoice, the field _Vehicle.total_expense_ must be updated automatically.

**Comment by JPS09** [comment]
> 
> 
> One important feature is missing.
> When you add or edit an invoice, the field _Vehicle.total_expense_ must be updated automatically.

I added the feature, feel free to have a look and check on your side if it works properly. I checked the main issue and it seems to be working fine on my side but I may have missed something

File: app/controllers/invoices_controller.rb [hunk 1]
```diff
@@ -1,7 +1,73 @@
...
+    @action = 'Ajouter'
+  end
+
+  def create
+    @invoice = Invoice.new(invoice_params)
+    @invoice.vehicle = @vehicle
+    @vehicle.total_expenses += params[:invoice][:price].to_f
+
+    if @invoice.save
+      redirect_to vehicle_invoice_path(@vehicle, @invoice), notice: 'Facture créee'
```
Comment (sbriclot) [inline app/controllers/invoices_controller.rb:24]: _total_expenses_ calculation is not at the right place. If the invoice is invalid, you will call the method **create** twice (or more), and the total will be increased each time the method is called.
Please move line 21 inside the _IF_ in order to update the total only when the invoice is valid ans saved.

Same change to do in the method **update**, move the call to _old_invoice_price_ inside the _IF_.

**Review by sbriclot — CHANGES_REQUESTED** [review]

**Comment by sbriclot** [comment]
If you received a mail about my code review, don't take care of it for the moment, I have a doubt about the position of total_expenses calculation, but I want to test your branch first, I'm not sure how rails manage commit and rollback in this case.

**Comment by JPS09** [comment]
> If you received a mail about my code review, don't take care of it for the moment, I have a doubt about the position of total_expenses calculation, but I want to test your branch first, I'm not sure how rails manage commit and rollback in this case.

Don't worry, I didn't received a notification.
Feel free to suggest a better way, I think that mine might leave some edge cases unattended but I can't pinpoint which ones

For trying commits, I think that if you checkout to the origin/my branch, you are in a kind of experimental mode in which your modification won't affect the pr, your terminal should give you a message when you enter this mode

File: app/controllers/invoices_controller.rb [hunk 2]
```diff
@@ -1,7 +1,73 @@
...
+
+  def old_invoice_price
+    @old_price = @invoice.price
+    @new_price = params[:invoice][:price].to_f
+
+    return unless @old_price != @new_price
+
+    @vehicle.total_expenses -= @old_price
+    @vehicle.total_expenses += @new_price
   end
```
Comment (sbriclot) [inline app/controllers/invoices_controller.rb]:
```suggestion
  def update_total_expenses(old_price)
    return if old_price == @invoice.price

    @vehicle.total_expenses += @invoice.price - old_price
    @vehicle.save
  end
```

**Review by sbriclot — CHANGES_REQUESTED** [review]

File: app/controllers/invoices_controller.rb (hunk 1 above)
Comment (sbriclot) [inline app/controllers/invoices_controller.rb]:
```suggestion
  def create
    @invoice = Invoice.new(invoice_params)
    @invoice.vehicle = @vehicle
    if @invoice.save
      update_total_expenses(0)
      redirect_to vehicle_invoice_path(@vehicle, @invoice), notice: 'Facture créee'
```

**Review by sbriclot — CHANGES_REQUESTED** [review]

File: app/controllers/invoices_controller.rb [hunk 3]
```diff
@@ -1,7 +1,73 @@
...
+  def edit
+    @action = 'Modifier'
+  end
+
+  def update
+    # Si il y a modification du prix, soustrait l'ancien prix et le remplace par le nouveau
+    old_invoice_price
+
+    if @invoice.update(invoice_params)
+      redirect_to vehicle_invoice_path(@vehicle, @invoice), notice: 'Facture mise à jour'
```
Comment (sbriclot) [inline app/controllers/invoices_controller.rb]:
```suggestion
  def update
    # Si il y a modification du prix, soustrait l'ancien prix et le remplace par le nouveau
    old_price = @invoice.price
    if @invoice.update(invoice_params)
      update_total_expenses(old_price)
      redirect_to vehicle_invoice_path(@vehicle, @invoice), notice: 'Facture mise à jour'
```

**Review by sbriclot — CHANGES_REQUESTED** [review]

**Comment by sbriclot** [comment]
**total_expenses** is not calculated at the right place and the change is not saved due to the missing command @vehicle.save.
Please have a look at the proposals for :
* the new name for the method **old_invoice_price**, which become **update_total_expenses**
* the method **update_total_expenses**, which can be used for invoice creation and update
* the calls to this method in **create** and **update**

**Comment by JPS09** [comment]
> 
> 
> **total_expenses** is not calculated at the right place and the change is not saved due to the missing command @vehicle.save.
> Please have a look at the proposals for :
> 
>     * the new name for the method **old_invoice_price**, which become **update_total_expenses**
> 
>     * the method **update_total_expenses**, which can be used for invoice creation and update
> 
>     * the calls to this method in **create** and **update**

Thank you for your suggestion, I committed them to the branch. I will work on the layout before Monday if I can.

**Review by sbriclot — APPROVED** [review]

**Comment by sbriclot** [comment]
Good job, works fine now, can be merged.

**Comment by JPS09** [comment]
I will work on the layout after merging this one since bootstrap is still in my branch and frankly, is getting in my way.

**Comment by JPS09** [comment]
When you can, can you give this PR a last review before it can be merged? Thank you!

**Comment by sbriclot** [comment]
Works fine.
Maybe a last validation rule to add, to avoid dates in the future (for me it's not possible to add an invoice before its edition).
What do you think about this proposal ?

**Comment by JPS09** [comment]
> 
> 
> Works fine.
> Maybe a last validation rule to add, to avoid dates in the future (for me it's not possible to add an invoice before its edition).
> What do you think about this proposal ?

I totally agree, I just added a new validation to prevent this, please have a look :)

File: app/models/invoice.rb [hunk 4]
```diff
@@ -3,4 +3,18 @@ class Invoice < ApplicationRecord
...
+  validate :invoice_date_cannot_be_in_the_future
+  validates :km, presence: true, numericality: { greater_than: 0 }
+  validates :price, presence: true, numericality: { greater_than: 0 }
+  validates :invoice_type_id, presence: true
+  validates :shop_id, presence: true
+
+  def invoice_date_cannot_be_in_the_future
+    return unless date.present? && date.strftime > Date.today.strftime
+
+    errors.add(:date, "Une facture ne peut être créée depuis le future")
```
Comment (sbriclot) [inline app/models/invoice.rb]:
```suggestion
    errors.add(:date, "Une facture ne peut être créée dans le futur")
```

**Review by sbriclot — CHANGES_REQUESTED** [review]

**Review by sbriclot — APPROVED** [review]

**Pull request merged**